ntoskrnl/se/token.c

```diff
@@ -135,34 +135,147 @@ SepDeleteTokenLock(
     ExFreePoolWithTag(Token->TokenLock, TAG_SE_TOKEN_LOCK);
 }
 
-static NTSTATUS
-SepCompareTokens(IN PTOKEN FirstToken,
-                 IN PTOKEN SecondToken,
-                 OUT PBOOLEAN Equal)
+/**
+ * @brief
+ * Compares tokens if they're equal based on all the following properties. If all
+ * of the said conditions are met then the tokens are deemed as equal.
+ *
+ * - Every SID that is present in either token is also present in the other one.
+ * - Both or none of the tokens are restricted.
+ * - If both tokens are restricted, every SID that is restricted in either token is
+ *   also restricted in the other one.
+ * - Every privilege present in either token is also present in the other one.
+ * 
+ * @param[in] FirstToken
+ * The first token.
+ *
+ * @param[in] SecondToken
+ * The second token.
+ *
+ * @param[out] Equal
+ * The retrieved value which determines if the tokens are
+ * equal or not.
+ *
+ * @return
+ * Returns STATUS_SUCCESS.
+ */
+static
+NTSTATUS
+SepCompareTokens(
+    _In_ PTOKEN FirstToken,
+    _In_ PTOKEN SecondToken,
+    _Out_ PBOOLEAN Equal)
 {
     BOOLEAN Restricted, IsEqual = FALSE;
+    ULONG FirstTokenIndex, SecondTokenIndex;
+    PAGED_CODE();
 
     ASSERT(FirstToken != SecondToken);
 
     /* Lock the tokens */
     SepAcquireTokenLockShared(FirstToken);
     SepAcquireTokenLockShared(SecondToken);
 
-    /* FIXME: Check if every SID that is present in either token is also present in the other one */
+    /* Bail out if user & group counters are not equal */
+    if (FirstToken->UserAndGroupCount != SecondToken->UserAndGroupCount)
+    {
+        DPRINT1("SepCompareTokens(): User & Group index counters are not the same!\n");
+        goto Quit;
+    }
+
+    /* Check if every SID that is present in either token is also present in the other one */
+    for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->UserAndGroupCount; FirstTokenIndex++)
+    {
+        for (SecondTokenIndex = FirstTokenIndex; SecondTokenIndex < SecondToken->UserAndGroupCount; SecondTokenIndex++)
```

Comment:
OK, let me tell you why this is wrong with an example (struct members simplified for clarity)

Say Token is the following : 
`{{AdminGroup, Attribute1}, {AdminUser, Attribute2}}`
Now apply your double loop to this token:

1. Token[0].Sid == Token[0].Sid --> OK
2. Token[0].Sid != Token[1].Sid --> Quit Loop

And your token is not equal to itself.